### ISSUE TITLE
sm120 (Blackwell): sageattn produces silently-wrong output when a captured CUDA graph is replayed with new input values (no NaN)

### ISSUE THREAD

**Issue by keeganmccallum** [thu-ml/SageAttention#392]
## Summary

`sageattn_qk_int8_pv_fp8_cuda` (the sm120/Blackwell dispatch path from `sageattn(..., tensor_layout=...)`) appears to produce silently-wrong (not NaN) output when the op is captured inside a `torch.cuda.CUDAGraph` and then **replayed after the input tensors' static buffers are overwritten with new values** — i.e. the standard "capture once, replay every step with fresh data" pattern used by `torch.cuda.graph()` / `make_graphed_callables` for diffusion/inference loops.

Calling `sageattn(...)` eagerly (uncaptured) is correct. Replaying a captured graph with the **same** values it was captured/primed with is also correct (small, expected INT8/FP8 quantization noise). Replaying with **different** values written into the same static input buffers diverges by ~90% relative to an eager call on those same new values — no NaN/Inf, just wrong numbers, which made this hard to notice (a full downstream pipeline decoded to pure static/noise video before we isolated it here).

## Environment

- GPU: NVIDIA RTX PRO 6000 Blackwell (Workstation/Server Edition), compute capability (12, 0) → `sageattn`'s own dispatch resolves to `"sm120"` → `sageattn_qk_int8_pv_fp8_cuda(..., qk_quant_gran="per_warp", pv_accum_dtype="fp32+fp16")`.
- Built from source (`git clone` + `pip install --break-system-packages --no-build-isolation -e .`, `TORCH_CUDA_ARCH_LIST=12.0`), commit `d1a57a546c3d395b1ffcbeecc66d81db76f3b4b5` at time of testing (this issue is against current `main`, please let us know if this has since moved).
- torch 2.13.0+cu13.0, CUDA 13.0 toolkit, Python 3.12.3.
- Confirmed the `_qattn_sm89` extension module (which the `"sm120"` dispatch branch actually calls into, via `sm89_compile.*`) genuinely compiles native `sm_120a` code for this build (`setup.py`'s `NVCC_FLAGS` include `-gencode arch=compute_120a,code=sm_120a` and that flag list is shared into the `_qattn_sm89` `CUDAExtension` whenever `HAS_SM120` is set) — this is not a case of an SM89-only binary being silently substituted.

## Minimal repro

```python
import torch
from sageattention import sageattn

def sage2_nhd(q, k, v):
    return sageattn(q, k, v, tensor_layout="NHD")

device = torch.device("cuda")
B, T, HEADS, DIM_HEAD = 1, 2550, 32, 128
HD = HEADS * DIM_HEAD

def rand_qkv():
    return (
        torch.randn(B, T, HD, device=device, dtype=torch.bfloat16).view(B, T, HEADS, DIM_HEAD),
        torch.randn(B, T, HD, device=device, dtype=torch.bfloat16).view(B, T, HEADS, DIM_HEAD),
        torch.randn(B, T, HD, device=device, dtype=torch.bfloat16).view(B, T, HEADS, DIM_HEAD),
    )

q0, k0, v0 = rand_qkv()
q_buf, k_buf, v_buf = q0.clone(), k0.clone(), v0.clone()

# Warmup on a side stream (standard pre-capture pattern)
s = torch.cuda.Stream()
s.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(s):
    for _ in range(3):
        _ = sage2_nhd(q_buf, k_buf, v_buf)
torch.cuda.current_stream().wait_stream(s)

g = torch.cuda.CUDAGraph()
with torch.cuda.graph(g):
    static_out = sage2_nhd(q_buf, k_buf, v_buf)

# Replay 1: same values used to prime/capture -- sanity check
g.replay()
torch.cuda.synchronize()
out_replay_priming = static_out.clone()
out_eager_priming = sage2_nhd(q0, k0, v0)
d1 = (out_replay_priming.float() - out_eager_priming.float()).abs()
print("replay vs eager on PRIMING values:", 100 * d1.max().item() / out_eager_priming.float().abs().max().item(), "%")
# -> ~3% (expected INT8/FP8 quantization noise)

# Replay 2: overwrite the SAME static buffers with NEW values, replay again
q1, k1, v1 = rand_qkv()
q_buf.copy_(q1); k_buf.copy_(k1); v_buf.copy_(v1)
g.replay()
torch.cuda.synchronize()
out_replay_new = static_out.clone()
out_eager_new = sage2_nhd(q1, k1, v1)
d2 = (out_replay_new.float() - out_eager_new.float()).abs()
print("replay vs eager on NEW (post-priming) values:", 100 * d2.max().item() / out_eager_new.float().abs().max().item(), "%")
# -> ~92% (!) -- no NaN/Inf, just numerically wrong
print("NaN:", torch.isnan(out_replay_new).any().item(), "Inf:", torch.isinf(out_replay_new).any().item())
```

Output on our machine (this exact standalone form of the script was still
mid-verification when our GPU scratch pod's session TTL expired and reclaimed
it -- the 2.96%/91.6% numbers below are from the equivalent test we DID run
to completion and verified on real hardware, using our own capture-safe
`torch.library.custom_op` wrapper around `sageattn` with the same shapes/
params, not this exact unwrapped script -- flagging that distinction rather
than claiming a bit-for-bit rerun we don't have. The two forms should be
numerically equivalent -- same op, same shapes, same dispatch path -- but we
want to be precise about what was and wasn't independently re-verified):
```
replay vs eager on PRIMING values: 2.96 %
replay vs eager on NEW (post-priming) values: 91.6 %
NaN: False Inf: False
```

## What we ruled out before filing

- **Not a wrapper/integration bug on our side**: we wrap `sageattn` behind a `torch.library.custom_op` (for `torch.compile`/graph-capture safety generally) and verified the SAME divergence happens calling `sageattn` completely raw, with no wrapper at all, using exactly the repro above.
- **Not general Sage2 numerical accuracy**: a single uncaptured call (eager, real attention-shaped tensors from an actual diffusion transformer layer) matches `torch.nn.functional.scaled_dot_product_attention` within 3.6% relative max diff — normal/expected for an INT8-QK/FP8-PV approximate kernel.
- **Not error accumulation across many steps/layers**: a full 8-step, 48-layer diffusion denoising schedule using Sage2 attention throughout (bf16, no capture) decodes to a clean, coherent video, frame 1 through frame 33, no progressive degradation.
- **Not specific to our downstream 4-bit weight quantization**: the same "clean without capture, broken with capture+replay-on-new-values" pattern reproduces with or without our own separate NVFP4 weight quantization in the surrounding model, and (per the repro above) reproduces on `sageattn` completely standalone with no model at all.
- **Not a wrong-architecture-kernel substitution**: confirmed via `setup.py`'s own `NVCC_FLAGS` construction that the `"sm120"` dispatch path's `_qattn_sm89` extension is compiled with genuine `sm_120a` gencode when built with `TORCH_CUDA_ARCH_LIST=12.0`.

## Where we'd look next (not confirmed — we haven't read the `.cu` sources)

`sageattn_qk_int8_pv_fp8_cuda` → `per_warp_int8_cuda`/`per_thread_int8_triton` (Q/K INT8 quant+scale) and `per_channel_fp8` (V quant) → `sm89_compile.qk_int8_sv_f8_accum_f16_fuse_v_scale_attn_inst_buf` (the fused kernel actually invoked for our config: `pv_accum_dtype="fp32+fp16"`). The Python-level `per_warp_int8` source allocates fresh `torch.empty()` scale/output tensors every call (no Python-side caching), so if there's a frozen-value effect it's most likely inside the compiled kernel/quantization CUDA code itself, or in how one of those kernel launches participates in (or fails to participate in) the CUDA-graph-capture stream.

## Why we think this matters for others, not just us

We only found this because we happened to test replay-with-different-values specifically (most capture-correctness checks, including our own platform's pre-existing self-test, only replay with the SAME priming values — which passes cleanly, at ~3% divergence, hiding this entirely). Anyone using `sageattn` inside a `torch.cuda.graph()`-captured inference loop on sm120 with `pv_accum_dtype="fp32+fp16"` and per-step-varying inputs (i.e. the standard "capture once, replay every step" pattern) would hit this silently — the output has no NaN/Inf to signal that anything is wrong, it just looks like severely degraded/garbled generation output.

Happy to help test a fix or narrow this down further if useful — this is a from-source build on real hardware, not a packaging issue on our end as far as we can tell.

**Comment by keeganmccallum** [comment]
Update: independently re-ran this exact standalone script (unwrapped, no framework code) on a second, freshly-provisioned instance of the same hardware/software config. Reproduces cleanly:

```
replay vs eager on PRIMING values: 2.94 %
replay vs eager on NEW (post-priming) values: 119.3 %
NaN: False Inf: False
```

Consistent order of magnitude with the original run (2.96% / 91.6%). The divergence on fresh post-capture values is robust across separate builds (rebuilt from source both times, both against the same pinned commit d1a57a546c3d395b1ffcbeecc66d81db76f3b4b5 of this repo) and separate runs (different random seeds for the synthetic q/k/v each time). Dropping any uncertainty caveat about the repro script itself -- it reproduces as written.

**Comment by cuihangbin** [comment]
> 更新：在第二个全新配置的相同硬件/软件配置实例上，独立地重新运行了完全相同的独立脚本（未封装，无框架代码）。问题可以顺利复现：
> 
> ```
> replay vs eager on PRIMING values: 2.94 %
> replay vs eager on NEW (post-priming) values: 119.3 %
> NaN: False Inf: False
> ```
> 
> 与原始运行结果（2.96% / 91.6%）的数量级一致。捕获后新值的偏差在不同的构建版本（两次均从源代码重新构建，且均基于此仓库的同一个固定提交[d1a57a5](https://github.com/thu-ml/SageAttention/commit/d1a57a546c3d395b1ffcbeecc66d81db76f3b4b5)）和不同的运行版本（每次使用不同的合成 q/k/v 随机种子）中均表现稳定。无需对复现脚本本身抱有任何疑虑——它能够按预期复现问题。

是的我也遇到了，这个绝对地址是错误的，没有实时更新kv 
需要修改算子
